overlord/configstate/configcore/ctrlaltdel.go

```diff
@@ -0,0 +1,142 @@
+	validAction := map[string]bool{
+		ctrlAltDelReboot: true,
+		ctrlAltDelNone:   true,
+	}
+	if !validAction[action] {
+		return fmt.Errorf("invalid action %q supplied for system.ctrl-alt-del-action option", action)
+	}
+
+	// The opts argument tells us if the real rootfs is mounted and
+	// systemd is running normally.
+	// opts != nil: The direct rootfs path is supplied because the system is not
+	//              ready so we cannot use systemctl to modify unit files
+	//              without supplying the root path.
+	// opts == nil: No rootfs path is supplied because we are running with
+	//              rootfs mounted, so we can use systemctl normally. This
+	//              case is used for normal runtime changes.
+	var sysd systemd.Systemd
+	if opts != nil {
+		// Use systemctl for direct unit file manipulations (support a
+		// subset of unit operations such as enable/disable/mark/unmask)
+		// See: "systemctl --root=<dir>"
+		sysd = systemd.NewEmulationMode(opts.RootDir)
+	} else {
+		// Use systemctl with access to all the unit operations
+		sysd = systemd.New(systemd.SystemMode, &sysdCtrlAltDelLogger{})
+
+		// Make sure the ctrl-alt-del.target unit is in the expected state.
+		// (1) The required unit should be present (file exist under /lib/systemd/system).
+		// (2) The Enable state for a unit typically means automatic startup on boot. The
+		//     expected state for reboot.target (ctrl-alt-del.target) is 'disabled'.
+		//     The ctrl-alt-del.target unit is an alias for reboot.target. This means that
+		//     if reboot.target is enabled, a ctrl-alt-del.target symlink will be created
+		//     under /etc/systemd/system, which is not needed and will prevent masking
+		status, err := sysd.Status(ctrlAltDelTarget)
+		if err != nil {
+			return err
+		}
+		switch {
+		case len(status) != 1:
+			return fmt.Errorf("internal error: expected status of target %s, got %v", ctrlAltDelTarget, status)
+		case !status[0].Installed:
+			return fmt.Errorf("internal error: target %s not installed", ctrlAltDelTarget)
+		case status[0].Enabled:
+			return fmt.Errorf("internal error: target %s should not be enabled", ctrlAltDelTarget)
+		}
+	}
+
+	// It is safe to mask an already masked unit, and unmask an already unmasked unit. The code
+	// will not try to optimize this because this will require knowledge of the initial
+	// "unset" state, which complicates the problem unnecessarily, for not much benefit.
+	switch action {
+	case ctrlAltDelNone:
+		// Create and make the /etc/systemd/system/ctrl-alt-del.target -> /dev/null
+		// The /etc/system/system/* unit directory is searched before the installed
+		// unit file directory /lib/system/system/* (which contains the valid symlink
+		// ctrl-alt-del.target -> reboot.target in the same directory).
+		if err := sysd.Mask(ctrlAltDelTarget); err != nil {
+			return err
+		}
+	case ctrlAltDelReboot:
+		// Remove the symlink /etc/systemd/system/ctrl-alt-del.target -> /dev/null
```

Comment:
// the unit is no longer masked and thus can be started on demand causing a reboot